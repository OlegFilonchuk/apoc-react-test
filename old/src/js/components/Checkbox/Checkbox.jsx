import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './Checkbox.less';
import { getDataTestElementProps, dataTestElementPropTypes } from '../../utils/dataTestElementPropUtils';

export const labelPositions = {
  left: 'left',
  right: 'right'
};

let globalCheckboxIndex = 0;

/**
 * Checkbox Component
 */
export default class Checkbox extends React.Component {
  componentWillMount() {
    globalCheckboxIndex += 1;
    this.id = `checkbox-gid-${globalCheckboxIndex}`;
  }

  render() {
    const {
      label,
      labelPosition,
      fullWidth,
      disabled,
      name,
      isOnlyCheckboxClickable,
      showHint,
      className,
      tooltip,
      ...props
    } = this.props;

    const id = name || this.id;

    const classes = classNames('square-checkbox', className, {
      'square-checkbox--clickable': !isOnlyCheckboxClickable,
      'square-checkbox-full-width': fullWidth,
      'square-checkbox-disabled': disabled
    });

    const labelTitle = showHint ? tooltip : null;

    const squareElement = (
      <label className="square-checkbox-square">
        <input
          type="checkbox"
          {...props}
          id={id}
          name={name}
          disabled={disabled}
          ref={ref => {
            this.input = ref;
          }}
          {...getDataTestElementProps(this.props)}
        />
        <span />
      </label>
    );

    const forElementId = !isOnlyCheckboxClickable ? id : '';
    const labelClasses = classNames('square-checkbox__label', {
      'square-checkbox__label--clickable': !isOnlyCheckboxClickable
    });
    const labelElement = (
      <label htmlFor={forElementId} className={labelClasses} title={labelTitle}>
        {label}
      </label>
    );

    const isLabelLeftPosition = labelPosition === labelPositions.left;
    const [firstItem, secondItem] = isLabelLeftPosition ? [labelElement, squareElement] : [squareElement, labelElement];

    return (
      <div className={classes}>
        {firstItem}
        {secondItem}
      </div>
    );
  }
}

Checkbox.propTypes = {
  /**
   * An additional class name that can be added to checkbox label for
   * extending it's functionality
   */
  className: PropTypes.string,

  /**
   * If set to yes label and checkbox fits the opposite edges of its container,
   * otherwise are positioned like two inline elements next to each other
   */
  fullWidth: PropTypes.bool,

  /**
   * Determines whether label should be positioned left from checkbox or right
   */
  labelPosition: PropTypes.oneOf([labelPositions.left, labelPositions.right]),

  /**
   * A flag to show title hint on hover
   */
  showHint: PropTypes.bool,

  /**
   * Label text or element.
   */
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,

  /**
   * This value will be shown as a hint on mouse over
   */
  tooltip: PropTypes.string,

  /**
   * Whether checkbox is disabled or not
   */
  disabled: PropTypes.bool,

  /**
   * It's passed to `input` element. Used also to create an `id` attribute of input
   * and `for` of label in order to connect those two.
   */
  name: PropTypes.string,

  /**
   * If is set to true label will not be clickable
   */
  isOnlyCheckboxClickable: PropTypes.bool,

  /**
   * Is the checkbox checked
   */
  checked: PropTypes.bool,
  ...dataTestElementPropTypes
};

Checkbox.defaultProps = {
  className: '',
  tooltip: '',
  name: '',
  disabled: false,
  fullWidth: false,
  showHint: false,
  isOnlyCheckboxClickable: false,
  labelPosition: labelPositions.right,
  checked: false
};
