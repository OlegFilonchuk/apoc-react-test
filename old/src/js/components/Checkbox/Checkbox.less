@disabledColor: #aeaeae;
@checkedColor: #555555;
@hoverColor: #aeaeae;

.square-checkbox {
  color: @checkedColor;
  display: flex;
  align-items: center;
  justify-content: flex-start;

  &__label {
    margin-bottom: 0;
    flex: 1 1;

    &--clickable {
      cursor: pointer;
    }
  }

  @squareSize: 12px;
  &-square {
    flex: 0 0 @squareSize;
    vertical-align: middle;
    display: inline-block;
    width: @squareSize;
    height: @squareSize;
    border: 1px solid @checkedColor;
    background-color: transparent;
    position: relative;
    margin: 0;

    input[type='checkbox'] {
      display: none;

      & + span {
        background-color: transparent;
        position: absolute;
        top: 1px;
        left: 1px;
        right: 1px;
        bottom: 1px;
        cursor: pointer;
        margin-bottom: 0;
      }
    }

    input[type='checkbox'][disabled]:not(:checked) + span {
      background-color: transparent;
    }

    input[type='checkbox'][disabled]:checked + span {
      background-color: @disabledColor;
    }

    input[type='checkbox']:checked:not([disabled]) + span {
      background-color: @checkedColor;
    }

    &:hover {
      input[type='checkbox']:not(:checked):not([disabled]) + span {
        background-color: @hoverColor;
      }
    }
  }

  &--clickable {
    &:hover {
      .square-checkbox-square {
        input[type='checkbox']:not(:checked):not([disabled]) + span {
          background-color: @hoverColor;
        }
      }
    }
  }

  &__label + &-square {
    margin-left: 8px;
  }

  &-square + &__label {
    margin-left: 8px;
  }
}

.square-checkbox.square-checkbox-disabled {
  color: @disabledColor;

  .square-checkbox-square {
    border-color: @disabledColor;
    cursor: not-allowed;

    > span {
      cursor: inherit;
    }
  }

  .square-checkbox__label.square-checkbox__label--clickable {
    cursor: not-allowed;
  }
}

.square-checkbox.square-checkbox-full-width {
  justify-content: space-between;

  .square-checkbox__label,
  .square-checkbox-square {
    margin-left: 0;
  }

  .square-checkbox-square:first-child + .square-checkbox__label,
  .square-checkbox__label:first-child + .square-checkbox-square {
    margin-left: auto;
  }
}
