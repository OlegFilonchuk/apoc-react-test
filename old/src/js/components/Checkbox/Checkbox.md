## Chameleon input\[type="checkbox"\]
This is example of checkbox which is just static html and css from Chameleon.

    <div className="sc-checkbox">
      <label>
        <input type="checkbox"/>
        Check me out
      </label>
    </div>

## Checkbox
Following are examples of Checkbox react component, which is in fact
a wrapper of html element:

```html
<input type="checkbox"/>
```

with no changes in functionality comparing to original `input`.
All non-Checkbox specific props passed to `Checkbox` are passed to `input`.

### Label position

By default, label position is to the right

    <div>
      <Checkbox label="Lorem ipsum"/>
    </div>

### Label position to the left
Position can be changed using `labelPosition` prop. All possible
values are stored in `labelPositions` map. It can be imported in following
way:

```html
import Checkbox, {labelPositions} from 'apoc-react/Checkbox';

```

    const labelPositions = require('./Checkbox').labelPositions;

    <Checkbox label="Lorem ipsum" labelPosition={labelPositions.left}/>

### Disabled

    <div>
      <Checkbox label="Lorem ipsum" disabled/>
    </div>

### Full width

    <div>
      <Checkbox label="Lorem ipsum" fullWidth/>
    </div>

### Checked by default

    <div>
      <Checkbox label="Lorem ipsum" defaultChecked/>
    </div>

### Passing additional attributes
Specific to `input` element

    const onCheckboxChange = e => {
      setState({
        checkboxValue: `${e.target.name}:${e.target.value}`,
        checkboxChecked: `${e.target.checked}`
      });
    };

    (
      <div>
        <p>Current value is: {state.checkboxValue}</p>
        <p>Is checked: {state.checkboxChecked}</p>
        <Checkbox name={'foo'} onChange={onCheckboxChange} label={'Click me!'} value="IamSelected" />
      </div>
    )

### Passing element as a title
The `title` prop can accept element instead of string,

    <div>
      <Checkbox
        label={(
          <span>
            <b>Strong title</b>{'\u00A0'}
            <i>italic part</i>{'\u00A0'}
            <u>part with underline</u>
          </span>
        )}
      />
    </div>

### Bi-directional flow of checked state. Changing state from the outside of Checkbox component.
The `title` prop can accept element instead of string,

    const initialState = { isChecked : false }

    function toggleCheckbox(){
      setState({isChecked: !state.isChecked});
    };

    function onCheckboxChange(e){
      setState({isChecked: e.target.checked});
    };

    (
      <div>
        <Checkbox
          label="This is checkbox"
          checked={state.isChecked}
          onChange={onCheckboxChange}
        />
        <br/>
        <p><span>Is checked?: {String(state.isChecked)}</span></p>
        <br/>
        <button onClick={toggleCheckbox}>Toggle checkbox</button>
      </div>
    )

### With title hint
    <div>
      <Checkbox label="Lorem ipsum" showHint={true} tooltip="some example hint" />
    </div>

### Only checkbox clickable example

By default, label position is to the right

    <div>
      <Checkbox label="Lorem ipsum" isOnlyCheckboxClickable/>
    </div>
