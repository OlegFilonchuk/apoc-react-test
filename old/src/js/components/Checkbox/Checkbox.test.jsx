import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Checkbox, { labelPositions } from './Checkbox';

describe('<Checkbox />', () => {
  it('renders input[type="checkbox"] wrapped in html structure', () => {
    const checkbox = shallow(<Checkbox label={'adasdas'} />);

    expect(checkbox).toBePresent();

    const squareCheckbox = checkbox.find('.square-checkbox');

    expect(squareCheckbox).toBePresent();

    const square = squareCheckbox.find('.square-checkbox-square');
    const label = squareCheckbox.find('.square-checkbox__label');

    expect(square).toBePresent();
    expect(label).toBePresent();

    const realCheckbox = square.find('input[type="checkbox"]');
    const spanForDrawingCheckbox = square.find('span');

    expect(realCheckbox).toBePresent();
    expect(spanForDrawingCheckbox).toBePresent();
  });

  it('renders with label positioned to right from checkbox by default', () => {
    const checkbox = shallow(<Checkbox label={'asadasd'} />);

    expect(
      checkbox
        .find('.square-checkbox')
        .children()
        .first()
    ).toHaveClassName('square-checkbox-square');
    expect(
      checkbox
        .find('.square-checkbox')
        .children()
        .last()
    ).toHaveClassName('square-checkbox__label');
  });

  it('prop labelPosition controls position correctly', () => {
    const checkboxWithLabelToLeft = shallow(<Checkbox label={'adasd'} labelPosition={labelPositions.left} />);

    expect(
      checkboxWithLabelToLeft
        .find('.square-checkbox')
        .children()
        .first()
    ).toHaveClassName('square-checkbox__label');
    expect(
      checkboxWithLabelToLeft
        .find('.square-checkbox')
        .children()
        .last()
    ).toHaveClassName('square-checkbox-square');

    const checkboxWithLabelToRight = shallow(<Checkbox label={'asdasd'} labelPosition={labelPositions.right} />);

    expect(
      checkboxWithLabelToRight
        .find('.square-checkbox')
        .children()
        .first()
    ).toHaveClassName('square-checkbox-square');
    expect(
      checkboxWithLabelToRight
        .find('.square-checkbox')
        .children()
        .last()
    ).toHaveClassName('square-checkbox__label');
  });

  it('prop fullWidth=true adds css class', () => {
    const checkboxWithDefaultFullWidthValue = shallow(<Checkbox label={'asdasds'} />);
    const notFullWidthCheckbox = shallow(<Checkbox label={'asdasd'} fullWidth={false} />);

    expect(checkboxWithDefaultFullWidthValue.find('.square-checkbox')).not.toHaveClassName(
      'square-checkbox-full-width'
    );
    expect(notFullWidthCheckbox.find('.square-checkbox')).not.toHaveClassName('square-checkbox-full-width');

    const checkbox = shallow(<Checkbox label={'asdads'} fullWidth />);

    expect(checkbox.find('.square-checkbox')).toHaveClassName('square-checkbox-full-width');
  });

  it('prop label adds renders text', () => {
    const checkbox = shallow(<Checkbox label={'Abracadabra!'} />);

    expect(checkbox.find('.square-checkbox__label')).toHaveText('Abracadabra!');
  });

  it('prop label adds renders element', () => {
    const checkbox = shallow(<Checkbox label={<span>text in span</span>} />);

    const elementInLabel = checkbox.find('.square-checkbox__label').children('span');

    expect(elementInLabel.length).toEqual(1);
    expect(elementInLabel).toHaveText('text in span');
  });

  describe('when prop disabled is set', () => {
    describe('to `true`', () => {
      it('renders disabled component', () => {
        const wrapper = mount(<Checkbox label="test" disabled />);

        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });
    describe('to `false`', () => {
      it('renders enabled component', () => {
        const wrapper = mount(<Checkbox label="test" disabled={false} />);

        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });
  });

  describe('when onChange event is triggered on <input/>', () => {
    it('should have onChange callback called once', () => {
      const onChangeCallback = jest.fn();

      const checkbox = mount(<Checkbox label={<span>text in span</span>} onChange={onChangeCallback} />);

      expect(checkbox.find('input')).toBePresent();
      checkbox.find('input').simulate('change', {
        // we are simulating the real values from the browser
        target: {
          name: 'some-checkbox', // browser will call the event with target.name being the 'name' attribute
          checked: true // and here with target.checked being true/false whether the checkbox is checked or not
        }
      });

      expect(onChangeCallback).toHaveBeenCalledTimes(1);
      expect(onChangeCallback).toHaveBeenCalledWith(
        jasmine.objectContaining({
          target: {
            name: 'some-checkbox',
            checked: true
          }
        })
      );
    });
  });

  describe('alt text', () => {
    const label = 'some.label.content';
    const title = 'some.title.content';

    let checkBox;

    function mountCheckbox(showHint) {
      checkBox = shallow(<Checkbox label={label} tooltip={title} showHint={showHint} />);
    }

    const testLabeltTitle = expectedValue => {
      const labelElement = checkBox.find('.square-checkbox__label');
      const result = labelElement.props().title;

      expect(result).toEqual(expectedValue);
    };

    it('should have correct HTML structure', () => {
      const showHint = true;

      mount(<Checkbox label={label} tooltip={title} showHint={showHint} />);

      expect(checkBox).toMatchSnapshot();
    });

    it('should add a title on label element', () => {
      const showHint = true;

      // add something
      mountCheckbox(showHint);

      testLabeltTitle(title);
    });

    it('should not add a title on label element', () => {
      const showHint = false;

      mountCheckbox(showHint);

      testLabeltTitle(null);
    });
  });

  describe('className', () => {
    const className = 'some-class-name';

    let checkBox;

    function mountCheckbox(additionalClassName) {
      checkBox = shallow(<Checkbox label="some.label.content" className={additionalClassName} />);
    }

    const testClassNamePresence = expectedClassNamePresence => {
      const classNameIsPresent = checkBox.find(`.${className}`).length !== 0;

      expect(classNameIsPresent).toEqual(expectedClassNamePresence);
    };

    it('should add an additional classname for label', () => {
      mountCheckbox(className);

      testClassNamePresence(true);
    });

    it('should not add an additional classname for label', () => {
      mountCheckbox('');

      testClassNamePresence(false);
    });
  });
});
