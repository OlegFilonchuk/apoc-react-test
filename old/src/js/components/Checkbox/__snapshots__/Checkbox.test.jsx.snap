// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> alt text should have correct HTML structure 1`] = `undefined`;

exports[`<Checkbox /> when prop disabled is set to \`false\` renders enabled component 1`] = `
<Checkbox
  checked={false}
  className=""
  disabled={false}
  fullWidth={false}
  isOnlyCheckboxClickable={false}
  label="test"
  labelPosition="right"
  name=""
  showHint={false}
  tooltip=""
>
  <div
    className="square-checkbox square-checkbox--clickable"
  >
    <label
      className="square-checkbox-square"
    >
      <input
        checked={false}
        disabled={false}
        id="checkbox-gid-11"
        name=""
        type="checkbox"
      />
      <span />
    </label>
    <label
      className="square-checkbox__label square-checkbox__label--clickable"
      htmlFor="checkbox-gid-11"
      title={null}
    >
      test
    </label>
  </div>
</Checkbox>
`;

exports[`<Checkbox /> when prop disabled is set to \`true\` renders disabled component 1`] = `
<Checkbox
  checked={false}
  className=""
  disabled={true}
  fullWidth={false}
  isOnlyCheckboxClickable={false}
  label="test"
  labelPosition="right"
  name=""
  showHint={false}
  tooltip=""
>
  <div
    className="square-checkbox square-checkbox--clickable square-checkbox-disabled"
  >
    <label
      className="square-checkbox-square"
    >
      <input
        checked={false}
        disabled={true}
        id="checkbox-gid-10"
        name=""
        type="checkbox"
      />
      <span />
    </label>
    <label
      className="square-checkbox__label square-checkbox__label--clickable"
      htmlFor="checkbox-gid-10"
      title={null}
    >
      test
    </label>
  </div>
</Checkbox>
`;
